src/Language/Haskell/HGrep/Print.hs

```diff
@@ -31,12 +33,41 @@ printParseError (ParseError (loc, msg)) =
 
 printSearchResult :: PrintOpts -> SearchResult -> [Char]
 printSearchResult (PrintOpts co) (SearchResult anns ast) =
-  let src = EP.exactPrint ast anns in
+  -- Get the start position of the comment before search result
+  let annsPairs      = Map.toList anns
+      targetAnnPairs = L.filter (isSameLoc .fst) annsPairs
+      resLoc         = getSpanStartLine resSpan
+      startLineNum   =
+        case targetAnnPairs of
+          []             -> resLoc
+          ((_, ann) : _) ->
+            case EP.annPriorComments ann of
+              []                 -> resLoc
+              ((comment, _) : _) -> getSpanStartLine $ EP.commentIdentifier comment in
+  -- ignore empty lines before the actual result
+  let wholeSrc    = EP.exactPrint ast anns
+      (nill, src) = L.span null $ L.lines wholeSrc in
     case co of
       DefaultColours ->
-        hscolour src
+        L.unlines $ nill <> L.zipWith printLine [startLineNum..] src
       NoColours ->
-        src
+        wholeSrc
+  where
+    resSpan :: SrcLoc.SrcSpan
+    resSpan = SrcLoc.getLoc ast
+
+    isSameLoc :: EP.AnnKey -> Bool
+    isSameLoc (EP.AnnKey loc _) = loc == resSpan
+
+    printLine :: Int -> [Char] -> [Char]
+    printLine i l = show i <> "  " <> hscolour l
+
+    getSpanStartLine :: SrcLoc.SrcSpan -> Int
+    getSpanStartLine someSpan =
+      case SrcLoc.srcSpanStart someSpan of
+        SrcLoc.RealSrcLoc x -> SrcLoc.srcLocLine x
+        -- TODO: Don't know how to get rid of error
+        _                   -> error "SrcLoc is FastString"
```

Comment:
Yeah, we're going to need to get rid of this `error` call before merging. I don't think this case happens very often, but if it does we shouldn't crash.

The constructor in question - [UnhelpfulLoc](https://github.com/ghc/ghc/blob/429fafb5b2b1ac02fd04d9a98e30b5991125692c/compiler/basicTypes/SrcLoc.hs#L116) - looks like it is only used for Template Haskell / GHCi definitions. This makes me think we could ignore it pretty safely, since any code we see will definitely have been parsed.

My suggestion: Make `startLineNum` and `getSpanStartLine` return `Maybe Int`. When printing the result, if it's `Nothing`, don't prepend locations.